### ISSUE TITLE
Error in guided particle filtering for pre-implemented stochastic volatility model.

### ISSUE THREAD

**Issue by vwiela** [nchopin/particles#57]
I am currently trying to implement a guided filter for a SDE-model. In order to do so I try to understand more how one needs to set the proposal distribution. So far I understood that it gets as an input $xp$, which is an $(N,d)$ shaped array with $N$ being the number of particles and $d$ the dimension of the state space. And the output should be a distribution of $X_T|X_{t-1}$. 
To better understand this I tried to executed the pre-implemented stochastic volatility model and got several questions and one unexplainable error message:
The code I tried to run is:
```
# standard libraries
from matplotlib import pyplot as plt
import numpy as np
import seaborn as sb
import pandas as pd
import scipy
import scipy.stats as stats
from scipy import linalg

from collections import OrderedDict

# modules from particles
import particles  # core module
from particles import distributions as dists  # where probability distributions are defined
from particles import state_space_models as ssm  # where state-space models are defined
from particles.collectors import Moments
from particles import mcmc
from particles import kalman


class StochVol(ssm.StateSpaceModel):
    r"""Univariate stochastic volatility model.
    .. math::
        X_0 & \sim N(\mu, \sigma^2/(1-\rho^2)) \\
        X_t & = \mu + \rho(X_{t-1}-\mu) + \sigma U_t, \quad U_t\sim N(0,1) \\
        Y_t|X_t=x_t & \sim N(0, e^{x_t}) \\
    """
    default_params = {'mu': -1.02, 'rho': 0.9702, 'sigma': .178}
    # values taken from Pitt & Shephard (1999)

    def sig0(self):
        """std of X_0"""
        return self.sigma / np.sqrt(1. - self.rho**2)

    def PX0(self):
        return dists.Normal(loc=self.mu, scale=self.sig0())

    def EXt(self, xp):
        """compute E[x_t|x_{t-1}]"""
        return (1. - self.rho) * self.mu + self.rho * xp

    def PX(self, t, xp):
        return dists.Normal(loc=self.EXt(xp), scale=self.sigma)

    def PY(self, t, xp, x):
        return dists.Normal(loc=0., scale=np.exp(0.5 * x))

    def _xhat(self, xst, sig, yt):
        return xst + 0.5 * sig**2 * (yt**2 * np.exp(-xst) - 1.)

    def proposal0(self, data):
        # Pitt & Shephard
        return dists.Normal(loc=self._xhat(0., self.sig0(), data[0]),
                            scale=self.sig0())

    def proposal(self, t, xp, data):
        # Pitt & Shephard
        return dists.Normal(loc=self._xhat(self.EXt(xp),
                                           self.sigma, data[t]),
                            scale=self.sigma)

fk_guided = ssm.GuidedPF(ssm=StochVol)
```
And I got the error:
```
TypeError                                 Traceback (most recent call last)
/home/vinc777/masterthesis/two_variant_model/particles/guided_filter.ipynb Cell 29 in <module>
----> [1](vscode-notebook-cell://wsl%2Bubuntu/home/vinc777/masterthesis/two_variant_model/particles/guided_filter.ipynb#X40sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0) fk_guided = ssm.GuidedPF(ssm=StochVol)

File ~/src/particles/particles/state_space_models.py:321, in Bootstrap.__init__(self, ssm, data)
    319 self.ssm = ssm
    320 self.data = data
--> 321 self.du = self.ssm.PX0().dim

TypeError: PX0() missing 1 required positional argument: 'self'
```
This error seems onlogic for me and I don`t know how to resolve it, since normally the `self` is not an argument that needs to be passed.

Now the other questions:

The distribution, which is returned by `proposal`, should just generates one particle $X_t^i$ given its ancestor $X_{t-1}^i$ correct? What I don't understand is where the transition from the $(N,d)$ shaped input $xp$ to the just $d$-dimensional $xp^i$ happens and how then the $i$-th particle is generated exactly to its ancestor. Does this mean that the mean and cov given to the distribution are $(N,d)$ resp. $(N,d,d)$ shaped or just $d$ or $d\times d$ dimensional?

Thanks!

**Comment by AdrienCorenflos** [comment]
Hi,
@nchopin feel free to correct me

You did not instanciate the class in:
```python
fk_guided = ssm.GuidedPF(ssm=StochVol)
```

Should be 
```python
fk_guided = ssm.GuidedPF(ssm=StochVol())  # with your params
```

Regarding the second question, the distributions need to be batched along the first dimension (or if you prefer, vectorised).
If the scale of the proposal (it's not the case here) depended on the particle, then your last point would be correct indeed.

**Comment by vwiela** [comment]
Thanks for pointing to my mistake!

Ok so if I understand you correctly. If I don`'t want the scale to be dependent on the particle before, then the input shapes are $(N,d)$ for the mean and $(d,d)$ for the scale. 
If I want it to be dependent than the scale needs to have shape $(N,d,d)$ correct?

I don't exactly know how the `np.stats.normal` behaves if one gives as an input $(N,d)$ and $(N,d,d)$ shaped arrays, but is the call `proposal(t, xp, data).rvs(Size=xp.shape[0])` then really creating an array of $N$ random variables, where the first has mean$=x_mean[0]$ and scale=$x_scale[0]$ and so on, as we want it? 
SO basically do you know what the shape of `proposal(t, xp, data).rvs` will be if we have the particle dependent scale?

**Comment by AdrienCorenflos** [comment]
It seems you will want to use the multivariate normal distribution (if your d > 1).
Also I am not convinced  that @nchopin implemented the MVN distribution for batched covariances though, so you may need to do implement that yourself.

If you are worried, you can always test the distribution indepently of the PF.

**Comment by vwiela** [comment]
Okay, thanks for your help.
I Will try to implement an appropriate multivariate normal distribution it on my own then.
